### PR TITLE
adding TypeLoadException handling

### PR DESCRIPTION
**Pull request by alex-kondrashov**
These fixes relate to this issue #1300.

Basically I corrected all the places that I've found. The main idea behind those fixes is
 wrapping Type.GetType statement with try-catch and handling of TypeLoadException

There is only one commit with all the changes included.

### CODE REVIEW

**Comment by rogeralsing** [comment]
Looks good over all, I'm not sure if we should go with a generic error message such as "Could not find type ...." or if we should try to be more explicit per use case.

In the reference PR we do have a more descriptive message :
https://github.com/akkadotnet/akka.net/pull/1286/files#diff-e0a72019db08bc7dfe1a953689e00917R124

cc @akkadotnet/contributors

**Comment by Aaronontheweb** [comment]
I agree with @rogeralsing 

Could you make the error messages more explicit, @alex-kondrashov ? Like if the exception happens because we were unable to load a `Serializer` type could you change the error message to mention that we failed to load a Serializer? Does that make sense?

**Comment by Aaronontheweb** [comment]
@alex-kondrashov would you mind updating this?

**Comment by rogeralsing** [comment]
@Aaronontheweb can you provide some additional info on some of the cases here, it might be hard for an oudsider to know the exact details of each usecase, I'll see what I can fill in myself

**Comment by Aaronontheweb** [comment]
@rogeralsing yep, I'll help start putting together a list

File: src/core/Akka/Serialization/Serializer.cs [hunk 1]
```diff
@@ -121,7 +121,17 @@ public static int GetSerializerIdentifierFromConfig(Type type, ExtendedActorSyst
         {
```
Comment (rogeralsing) [inline src/core/Akka/Serialization/Serializer.cs]: cc @Horusiath what does this method actually do?
Each serializer does have an identifier on the serializer, why is this needed?
Can we not just use ` Sys.Serialization.FindSerializerFor(obj).Identifier`
Does the identifier need to be configurable?

Comment (Horusiath) [inline src/core/Akka/Serialization/Serializer.cs]: At some point Typesafe guys added ability to specify identifier for each serializer directly through config i.e. for case, when you have conflicting serializers.

File: src/core/Akka/IO/Dns.cs [hunk 2]
```diff
@@ -124,11 +124,19 @@ public DnsSettings(Config config)
...
-            Settings = new DnsSettings(system.Settings.Config.GetConfig("akka.io.dns"));
-            //TODO: system.dynamicAccess.getClassFor[DnsProvider](Settings.ProviderObjectName).get.newInstance()
-            Provider = (IDnsProvider) Activator.CreateInstance(Type.GetType(Settings.ProviderObjectName));
-            Cache = Provider.Cache;
+            try
+            {
+                _system = system;
+                Settings = new DnsSettings(system.Settings.Config.GetConfig("akka.io.dns"));
+                //TODO: system.dynamicAccess.getClassFor[DnsProvider](Settings.ProviderObjectName).get.newInstance()
+                Provider = (IDnsProvider)Activator.CreateInstance(Type.GetType(Settings.ProviderObjectName, true));
```
Comment (rogeralsing) [inline src/core/Akka/IO/Dns.cs]: IMO, `Type.GetType(Settings.ProviderObjectName, true)`should go on a separate line for clarity.
There should also be a check to ensure that the created object from `Activator.CreateInstance` actually is assignable to `IDnsProvder`

The message for the TypeLoadException should indicate that the type is to be used as a Dns Provider.
If the type is not of the correct type, there needs to be a `ConfigurationException` thrown explaining that the type is not a IDnsProvider.

File: src/core/Akka/Dispatch/Mailboxes.cs [hunk 3]
```diff
@@ -111,8 +120,16 @@ public Type FromConfig(string path)
...
-            var mailboxType = Type.GetType(type);
-            return mailboxType;
+            try
+            {
+                var mailboxType = Type.GetType(type, true);
+                return mailboxType;
+            }
+            catch (TypeLoadException ex)
+            {
+                var msgException = string.Format("Could not find type '{0}'", type);
```
Comment (rogeralsing) [inline src/core/Akka/Dispatch/Mailboxes.cs]: Here we have the full `path`, `path + ",mailbox-type"` so we can provide that in the type load exception.
"Could not find the mailbox type ${type} provided in the configuration path ${path + ".mailbox-path"}`"
We should also have a check that ensures that the type is assignable to `Akka.Dispatch.Mailbox`, if not, then we need to throw a configuration exception saying that the provided type is not a mailbox type

File: src/core/Akka/Actor/Internal/ActorSystemImpl.cs [hunk 4]
```diff
@@ -268,10 +276,18 @@ private void ConfigureMailboxes()
...
+            try
+            {
+                Type providerType = Type.GetType(_settings.ProviderClass, true);
+                global::System.Diagnostics.Debug.Assert(providerType != null, "providerType != null");
+                var provider = (IActorRefProvider)Activator.CreateInstance(providerType, _name, _settings, _eventStream);
+                _provider = provider;
+            }
+            catch (TypeLoadException ex)
+            {
+                var msgException = string.Format("Could not find type '{0}'", _settings.ProviderClass);
```
Comment (rogeralsing) [inline src/core/Akka/Actor/Internal/ActorSystemImpl.cs]: We should indicate that the type should be used as an `ActorRefProvider`..
e.g. "Could not find the actorref provider type '{0}'"
We also need to check that the type is assignable to `Akka.Actor.IActorRefProvider`, if not, throw a configuration exception indicating that the type is not a valid actor ref provider.

File: src/core/Akka/Actor/Deployer.cs [hunk 5]
```diff
@@ -100,11 +100,20 @@ private RouterConfig CreateRouterConfig(string routerTypeAlias, Config deploymen
...
             var routerTypeName = _settings.Config.GetString(path);
-            var routerType = Type.GetType(routerTypeName);
-            Debug.Assert(routerType != null, "routerType != null");
-            var routerConfig = (RouterConfig)Activator.CreateInstance(routerType, deployment);
 
-            return routerConfig;
+            try
+            {
+                var routerType = Type.GetType(routerTypeName, true);
+                Debug.Assert(routerType != null, "routerType != null");
```
Comment (rogeralsing) [inline src/core/Akka/Actor/Deployer.cs]: The debug.assert calls can be removed as they will never be hit as we are throwing before the line is executed.

File: src/core/Akka/Actor/Deployer.cs [hunk 6]
```diff
@@ -100,11 +100,20 @@ private RouterConfig CreateRouterConfig(string routerTypeAlias, Config deploymen
...
+                var routerConfig = (RouterConfig)Activator.CreateInstance(routerType, deployment);
+
+                return routerConfig;
+            }
+            catch (TypeLoadException ex)
+            {
+                var msgException = string.Format("Could not find type '{0}'", routerTypeName);
```
Comment (rogeralsing) [inline src/core/Akka/Actor/Deployer.cs]: We should indicate that the type is intended for a Rotuer Config.
e.g. "Could not find the router config type ${type}"

we should also throw an `ConfigurationException` if the type is not assignable to `Akka.Routing.RouterConfig`

File: src/core/Akka/Serialization/Serializer.cs (hunk 1 above)
Comment (rogeralsing) [inline src/core/Akka/Serialization/Serializer.cs]: Has this replaced the use of `serializer.Identifier` through out the Scala code? or are they assigning the value to the Identifier property using this?

(It seems to be a strange way to go about things as you might end up with one type of serializer on one node claiming to be identifier 1, and a completely different serializer with the same identifier on another node.)

Comment (Horusiath) [inline src/core/Akka/Serialization/Serializer.cs]: It isn't touching `Identifer` property, it can be only used to assign some value to it.

Comment (rogeralsing) [inline src/core/Akka/Serialization/Serializer.cs]: So ise the serializer.identifier obsolete and should be replaced with this new way of handling things?
It seems conflicting to have both? or maybe I'm just not understanding the difference :-)

Comment (Horusiath) [inline src/core/Akka/Serialization/Serializer.cs]: No. `Serializer.Identifier` is still valid. What only have changed, is that beside of having statically assigned number, now it can be a configurable value instead.

**Comment by alex-kondrashov** [comment]
I've fixed the messages in the cases mentioned in the list above. Please have a review on that.

File: src/core/Akka/Actor/Internal/ActorSystemImpl.cs [hunk 7]
```diff
@@ -132,8 +132,16 @@ public override ActorSelection ActorSelection(string actorPath)
 
         private void ConfigureScheduler()
         {
-            var schedulerType = Type.GetType(_settings.SchedulerClass, true);
-            _scheduler = (IScheduler) Activator.CreateInstance(schedulerType, this);
+            try
+            {
+                var schedulerType = Type.GetType(_settings.SchedulerClass, true);
+                _scheduler = (IScheduler)Activator.CreateInstance(schedulerType, this);
```
Comment (rogeralsing) [inline src/core/Akka/Actor/Internal/ActorSystemImpl.cs]: We need to check that the type is an IScheduler or throw configuration exception

File: src/core/Akka.Cluster/ClusterMetricsCollector.cs [hunk 8]
```diff
@@ -856,18 +856,16 @@ public static IMetricsCollector Get(ExtendedActorSystem system, ClusterSettings
...
-            if (metricsCollectorClass == null)
-            {
-                throw new ConfigurationException(string.Format("Could not create custom metrics collector {0}", fqcn));
-            }
-
             try
             {
-                var metricsCollector = (IMetricsCollector) Activator.CreateInstance(metricsCollectorClass, system);
+                var metricsCollectorClass = Type.GetType(fqcn, true);
+                var metricsCollector = (IMetricsCollector)Activator.CreateInstance(metricsCollectorClass, system);
```
Comment (rogeralsing) [inline src/core/Akka.Cluster/ClusterMetricsCollector.cs]: We should check that the type is a IMetricsCollector or throw configuration exception

File: src/core/Akka.Persistence/Journal/EventAdapters.cs [hunk 9]
```diff
@@ -164,13 +164,22 @@ private static EventAdapters Create(ExtendedActorSystem system, IDictionary<stri
...
+                    var type = Type.GetType(kv.Key, true);
+                    var adapter = kv.Value.Length == 1
+                            ? handlers[kv.Value[0]]
+                            : new CombinedReadEventAdapter(kv.Value.Select(h => handlers[h]));
+                    return new KeyValuePair<Type, IEventAdapter>(type, adapter);
+                }
+                catch (Exception ex)
+                {
+                    var msgException = string.Format("Could not find type '{0}'", kv.Key);
+                    throw new TypeLoadException(msgException, ex);
```
Comment (rogeralsing) [inline src/core/Akka.Persistence/Journal/EventAdapters.cs]: @Horusiath what would be the correct messages here?

File: src/core/Akka.Persistence/Persistence.cs [hunk 10]
```diff
@@ -156,17 +156,25 @@ private PluginHolder CreatePlugin(string configPath, Func<Type, string> dispatch
...
-            return new PluginHolder(pluginRef, CreateAdapters(configPath));
+            try
+            {
+                var pluginType = Type.GetType(pluginTypeName, true);
+                var shouldInjectConfig = pluginConfig.HasPath("inject-config") && pluginConfig.GetBoolean("inject-config");
+                var pluginDispatcherId = pluginConfig.HasPath("plugin-dispatcher")
+                    ? pluginConfig.GetString("plugin-dispatcher")
+                    : dispatcherSelector(pluginType);
+                var pluginActorArgs = shouldInjectConfig ? new object[] { pluginConfig } : null;
+                var pluginActorProps = new Props(pluginType, pluginActorArgs).WithDispatcher(pluginDispatcherId);
```
Comment (rogeralsing) [inline src/core/Akka.Persistence/Persistence.cs]: cc @Aaronontheweb @Horusiath should pluginType be checked for ActorBase compat here?

File: src/core/Akka.Remote/FailureDetectorRegistry.cs [hunk 11]
```diff
@@ -65,13 +65,16 @@ public static class FailureDetectorLoader
         /// <returns>A configured instance of the given <see cref="FailureDetector"/> implementation.</returns>
         public static FailureDetector Load(string fqcn, Config config, ActorSystem system)
         {
-            var failureDetectorClass = Type.GetType(fqcn);
-            if (failureDetectorClass == null)
+            try
+            {
+                var failureDetectorClass = Type.GetType(fqcn, true);
+                var failureDetector = (FailureDetector)Activator.CreateInstance(failureDetectorClass, config, system.EventStream);
```
Comment (rogeralsing) [inline src/core/Akka.Remote/FailureDetectorRegistry.cs]: Should check for FailureDetector compat or throw configuration exception

File: src/core/Akka.Remote/Transport/TransportAdapters.cs [hunk 12]
```diff
@@ -66,11 +66,16 @@ public TransportAdapters(ExtendedActorSystem system)
             {
                 try
                 {
-                    var adapterTypeName = Type.GetType(adapter.Value);
+                    var adapterTypeName = Type.GetType(adapter.Value, true);
                     // ReSharper disable once AssignNullToNotNullAttribute
                     var newAdapter = (ITransportAdapterProvider)Activator.CreateInstance(adapterTypeName);
```
Comment (rogeralsing) [inline src/core/Akka.Remote/Transport/TransportAdapters.cs]: should check for ITransportAdapterProvider compat or throw configuration exception

File: src/core/Akka/Serialization/Serializer.cs [hunk 13]
```diff
@@ -121,7 +121,17 @@ public static int GetSerializerIdentifierFromConfig(Type type, ExtendedActorSyst
...
-                .ToDictionary(pair => Type.GetType(pair.Key, true), pair => pair.Value.GetInt());
+                .ToDictionary(pair => {
+                    try
+                    {
+                        return Type.GetType(pair.Key, true);
+                    }
+                    catch (TypeLoadException ex)
+                    {
+                        var msgException = string.Format("Could not find serialization identifier type '{0}'", pair.Key);
+                        throw new TypeLoadException(msgException, ex);
```
Comment (rogeralsing) [inline src/core/Akka/Serialization/Serializer.cs]: @Horusiath is this the correct message?

File: src/core/Akka.Persistence/Persistence.cs (hunk 10 above)
Comment (Horusiath) [inline src/core/Akka.Persistence/Persistence.cs]: Yes, it should. But we could think of it as a separate issue.

File: src/core/Akka/Serialization/Serializer.cs (hunk 13 above)
Comment (Horusiath) [inline src/core/Akka/Serialization/Serializer.cs]: It's fine.

File: src/core/Akka.Persistence/Journal/EventAdapters.cs (hunk 9 above)
Comment (Horusiath) [inline src/core/Akka.Persistence/Journal/EventAdapters.cs]: Something like "Could not find event adapter of type '{0}'". It's more informative, but this one is still good.
